zcash_primitives/src/note_encryption.rs

```diff
@@ -0,0 +1,1079 @@
+
+        assert_eq!(
+            try_sapling_note_decryption(
+                &ivk,
+                &edwards::Point::<Bls12, _>::rand(&mut rng, &JUBJUB).mul_by_cofactor(&JUBJUB),
+                &cmu,
+                &enc_ciphertext
+            ),
+            None
+        );
+    }
+
+    #[test]
+    fn decryption_with_invalid_cmu() {
+        let mut rng = thread_rng();
+
+        let (_, ivk, _, _, epk, enc_ciphertext, _) = random_enc_ciphertext(&mut rng);
+
+        assert_eq!(
+            try_sapling_note_decryption(&ivk, &epk, &Fr::rand(&mut rng), &enc_ciphertext),
+            None
+        );
+    }
+
+    #[test]
+    fn decryption_with_invalid_tag() {
+        let mut rng = thread_rng();
+
+        let (_, ivk, _, cmu, epk, mut enc_ciphertext, _) = random_enc_ciphertext(&mut rng);
+
+        enc_ciphertext[ENC_CIPHERTEXT_SIZE - 1] ^= 0xff;
+        assert_eq!(
+            try_sapling_note_decryption(&ivk, &epk, &cmu, &enc_ciphertext),
+            None
+        );
+    }
+
+    #[test]
+    fn compact_decryption_with_invalid_ivk() {
+        let mut rng = thread_rng();
+
+        let (_, _, _, cmu, epk, enc_ciphertext, _) = random_enc_ciphertext(&mut rng);
+
+        assert_eq!(
+            try_sapling_compact_note_decryption(
+                &Fs::rand(&mut rng),
+                &epk,
+                &cmu,
+                &enc_ciphertext[..COMPACT_NOTE_SIZE]
+            ),
+            None
+        );
+    }
+
+    #[test]
+    fn compact_decryption_with_invalid_epk() {
+        let mut rng = thread_rng();
+
+        let (_, ivk, _, cmu, _, enc_ciphertext, _) = random_enc_ciphertext(&mut rng);
+
+        assert_eq!(
+            try_sapling_compact_note_decryption(
+                &ivk,
+                &edwards::Point::<Bls12, _>::rand(&mut rng, &JUBJUB).mul_by_cofactor(&JUBJUB),
+                &cmu,
+                &enc_ciphertext[..COMPACT_NOTE_SIZE]
+            ),
+            None
+        );
+    }
+
+    #[test]
+    fn compact_decryption_with_invalid_cmu() {
+        let mut rng = thread_rng();
+
+        let (_, ivk, _, _, epk, enc_ciphertext, _) = random_enc_ciphertext(&mut rng);
+
+        assert_eq!(
+            try_sapling_compact_note_decryption(
+                &ivk,
+                &epk,
+                &Fr::rand(&mut rng),
+                &enc_ciphertext[..COMPACT_NOTE_SIZE]
+            ),
+            None
+        );
+    }
+
+    #[test]
+    fn compact_decryption_with_invalid_diversifier() {
+        let mut rng = thread_rng();
+
+        let (_, ivk, _, cmu, epk, mut enc_ciphertext, _) = random_enc_ciphertext(&mut rng);
+
+        // In compact decryption, this will result in an altered diversifier
+        enc_ciphertext[1] ^= 0xff;
+        assert_eq!(
+            try_sapling_compact_note_decryption(
+                &ivk,
+                &epk,
+                &cmu,
+                &enc_ciphertext[..COMPACT_NOTE_SIZE]
+            ),
+            None
+        );
+    }
```

Comment:
There's a missing failure case for each of normal decryption and for recovery: the diversifier may be invalid so that g<sub>d</sub> can't be computed. `random_enc_ciphertext` will never create a ciphertext that triggers this case because it always uses a fixed (all-zeros) diversifier.

`compact_decryption_with_invalid_diversifier` covers only the case of the diversifier with (little-endian) value 0xFF. I don't know whether this is a valid diversifier or not, but either way it can only be covering one of the two possible cases: where the diversifier itself is invalid, or where it is valid but causes the commitment to be wrong.